frontend/src/components/table/ServerPaginatedDataTable.tsx

```diff
@@ -0,0 +1,198 @@
+import {
+  component$,
+  useStylesScoped$,
+  JSXChildren,
+  useSignal,
+  $,
+  QwikChangeEvent,
+  PropFunction,
+  useResource$,
+  Resource,
+} from '@builder.io/qwik';
+import { PaginationParams } from '~/types/paginated';
+import { FilterInput } from './DefaultFilter';
+import styles from './table.css?inline';
+
+export interface TableProps<T extends string> {
+  rows: Record<T, JSXChildren>[];
+  customColumnNames?: Partial<Record<T, string>>;
+  total: number;
+  emitFetchRows: PropFunction<
+    ({ page, limit, filter, sort, sortColumn }: PaginationParams) => Promise<{
+      total: number;
+      data: Record<T, JSXChildren>[];
+    }>
+  >;
+}
+
+export const ServerPaginatedDataTable = component$(<T extends string>(props: TableProps<T>) => {
+  useStylesScoped$(styles);
+  const rowsPerPage = useSignal(10);
+  const currentPage = useSignal(0);
+  const filter = useSignal('');
+  const sortColumn = useSignal<keyof T>(Object.keys(props.rows[0])[0] as keyof T);
+  const sortDesc = useSignal(true);
+  const maxPages = useSignal(props.total);
+
+  const headers = Object.keys(props.rows[0]).map((colName) => ({
+    name: colName,
+    displayName: props.customColumnNames?.[colName as T] || colName,
+  }));
+
+  const filterSetter$ = $((val: string) => {
+    filter.value = val;
+  });
+
+  const setSortParams = $((colName: keyof T) => {
+    sortColumn.value = colName;
+    sortDesc.value = !sortDesc.value;
+  });
+
+  const paginatedRows = useResource$(async ({ track }) => {
+    track(() => currentPage.value);
+    track(() => rowsPerPage.value);
+    track(() => filter.value);
+    track(() => sortDesc.value);
+    track(() => sortColumn.value);
+
+    const result = await props.emitFetchRows({
+      page: currentPage.value,
+      limit: rowsPerPage.value,
+      filter: filter.value,
+      sort: sortDesc.value ? 'desc' : 'asc',
+      sortColumn: sortColumn.value as string,
+    });
+
+    return result;
+  });
+
+  const isOnFirstPage = currentPage.value === 0;
+  const isOnLastPage = currentPage.value === maxPages.value - 1;
+
+  const setPageNumber = $((mode: 'prev' | 'next') => {
+    switch (mode) {
+      case 'prev':
```

Comment:
Consider use an enum

Reply:
its just one place...would benefit if used in other places